src/main/java/org/truffleruby/core/exception/CoreExceptions.java

```diff
@@ -161,9 +161,12 @@ public DynamicObject argumentErrorWrongArgumentType(Object object, String expect
     }
 
     @TruffleBoundary
-    public DynamicObject argumentErrorInvalidValue(Object object, String expectedType, Node currentNode) {
-        String badClassName = Layouts.MODULE.getFields(context.getCoreLibrary().getLogicalClass(object)).getName();
-        return argumentError(StringUtils.format("invalid value for %s(): %s", badClassName, expectedType), currentNode);
+    public DynamicObject argumentErrorInvalidStringToInteger(Object object, Node currentNode) {
+        assert RubyGuards.isRubyString(object);
+
+        // TODO (nirvdrum 19-Apr-18): Guard against String#inspect being redefined to return something other than a String.
+        final String formattedObject = StringOperations.getString((DynamicObject) context.send(object, "inspect"));
```

Comment:
It sounds like we should have a helper method to inspect a Ruby object, and deal with `#inspect` returning a non-String.
For example, RubyLanguage.toString() has similar logic.

Reply:
That would only work if you know the original object to inspect is a String (it's the case here but not in general).
I think it's fine to dynamically call `inspect` here.
It's possible to call an original method by capturing it, for instance by storing a Ruby UnboundMethod in a constant, but it needs a fair bit of boilerplate.